contracts/src/BeefyClient.sol

```diff
@@ -391,48 +425,42 @@ contract BeefyClient {
-     *
-     * One assumption is that Polkadot/Kusama will never have more than roughly 20,000 active validators in a session.
-     * As of writing this comment, Polkadot has 300 validators and Kusama has around 1000 validators,
-     * so we are well within those limits.
-     *
-     * In any case, an order of magnitude increase in validator set sizes will likely require a re-architecture
-     * of Polkadot that would make this contract obsolete well before the assumption becomes a problem.
-     *
-     * Constants generated with the help of scripts/minsigs.py
+     * See https://hackmd.io/9OedC7icR5m-in_moUZ_WQ for full analysis.
      */
-    function minimumSignatureThreshold(uint256 validatorSetLen) internal pure returns (uint256) {
-        if (validatorSetLen <= 10) {
-            return validatorSetLen - (validatorSetLen - 1) / 3;
-        } else if (validatorSetLen < 342) {
-            return 10;
-        } else if (validatorSetLen < 683) {
-            return 11;
-        } else if (validatorSetLen < 1366) {
-            return 12;
-        } else if (validatorSetLen < 2731) {
-            return 13;
-        } else if (validatorSetLen < 5462) {
-            return 14;
-        } else if (validatorSetLen < 10923) {
-            return 15;
-        } else if (validatorSetLen < 21846) {
-            return 16;
-        } else {
-            return 17;
+    function computeNumRequiredSignatures(
+        uint256 validatorSetLen,
+        uint256 signatureUsageCount,
+        uint256 minRequiredSignatures
+    ) internal pure returns (uint256) {
+        // Start with the minimum number of signatures.
+        uint256 numRequiredSignatures = minRequiredSignatures;
+
+        // We must substrate minimumSignatures from the number of validators or we might end up
+        // requiring more signatures than there are validators.
+        uint256 extraValidatorsLen = validatorSetLen.saturatingSub(minRequiredSignatures);
+        if (extraValidatorsLen > 0) {
+            numRequiredSignatures += Math.log2(extraValidatorsLen, Math.Rounding.Ceil);
         }
+
+        // To address the concurrency issue specified in the link below:
+        // https://hackmd.io/wsVcL0tZQA-Ks3b5KJilFQ?view#Solution-2-Signature-Checks-dynamically-depend-on-the-No-of-initial-Claims-per-session
+        // It must be harder for a malicious relayer to spam submitInitial to bias the RANDAO.
+        // If we detect that a signature is used many times (spam), we increase the number of signature samples required on submitFinal.
+        if (signatureUsageCount > 0) {
+            // Based on formula provided here: https://hackmd.io/9OedC7icR5m-in_moUZ_WQ
+            numRequiredSignatures += 1 + 2 * Math.log2(signatureUsageCount, Math.Rounding.Ceil);
+        }
+
+        // Never require more signatures than 2/3 majority.
+        uint256 validatorSetMajority = validatorSetLen - (validatorSetLen - 1) / 3;
```

Comment:
I believe the correct term is [quorum](https://en.wikipedia.org/wiki/Quorum_(distributed_computing)) instead of `majority`.

```suggestion
        uint256 validatorSetQuorum = validatorSetLen - (validatorSetLen - 1) / 3;
```

Also can we add unit test to ensure that `validatorSetLen - (validatorSetLen - 1) / 3` is always at least 2/3 of `validatorSetLen`. I don't think I tested that math previously.

Reply:
To clarify, I was wanting a test that covers `validatorSetLen - (validatorSetLen - 1) / 3` on its own.

Also that expression needs to be factored out into its function, as it is duplicated in `submitInitial`, where there is a check that the relayer provided a claims bitfield with at least 2/3 bits set.

Reply:
Addressed in https://github.com/Snowfork/snowbridge/pull/956/commits/d8144f3ca33f55bce887b03f0aea30ed736a5d8e